File: Week 3/id_520/102.二叉树的层次遍历.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode.cn id=102 lang=java
 *
 * [102] 二叉树的层次遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
    //     //方法一，DFS，迭代
    //     List<List<Integer>> result = new ArrayList<>();
    //     levelHelper(result, root, 0);
    //     return result;
    // }

    // public void levelHelper(List<List<Integer>> result, TreeNode root, int height) {
    //     //terminator
    //     if (root == null) return;

    //     //process current logic
    //     if (height >= result.size()) {
    //         result.add(new ArrayList<Integer>());
    //     }
    //     result.get(height).add(root.val);
    //     levelHelper(result, root.left, height+1);
    //     levelHelper(result, root.right, height+1);

    // }

    //方法二，迭代，用队列完成
    Queue<TreeNode> queue = new LinkedList<TreeNode>();
    List<List<Integer>> result = new ArrayList<>();
    
    if (root == null) return result;

    queue.offer(root);
    while(!queue.isEmpty()) {
        int levelNum = queue.size();
        List<Integer> subList = new ArrayList<>();
        for(int i = 0; i<levelNum; i++) {
            if(queue.peek().left != null) queue.offer(queue.peek().left);
            if(queue.peek().right != null) queue.offer(queue.peek().right);
            subList.add(queue.poll().val);
        }
        result.add(subList);
    }
    return result;
}
}
// @lc code=end

File: Week 3/id_445/102.二叉树的层次遍历.cpp
/*
 * @lc app=leetcode.cn id=102 lang=cpp
 *
 * [102] 二叉树的层次遍历
 */
#include <vector>
#include <queue>
using namespace std;
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
///solution 1 bfs
    

//      vector<vector<int>> levelOrder(TreeNode* root) {
//         vector<vector<int>> res;
//         queue<TreeNode *> q;
//         if(root)q.push(root);
//         int size=0;
        
//         TreeNode * n;
// ////good solution
//         // while (!q.empty())
//         // {
//         //     size=q.size();
//         //     vector<int> vec;
//         //     for (int i = 0; i < size; i++)
//         //     {
//         //         n=q.front();
//         //         vec.push_back(n->val);
//         //         q.pop();
//         //         if(n->left)q.push(n->left);
//         //         if(n->right)q.push(n->right);
//         //     }
//         //     res.push_back(vec);
            

//         // }

//         ///bad solution
//         vector<int> *vec=new vector<int>();
//         int i=0;
//         size=q.size();
//         while (i < size)
//         {
            
//             n=q.front();
//             vec->push_back(n->val);
//             q.pop();
//             if(n->left)q.push(n->left);
//             if(n->right)q.push(n->right);
//             i++;
//             if (vec->size()==size)
//             {
//                 size=q.size();
//                 i=0;
//                 res.push_back(*vec);
//                 vec=new vector<int> ;
//             }
            
//         }
        
        
//         return res;
//     }
    ///solution2 dfs
    vector<vector<int>> levelOrder(TreeNode* root) {
vector<vector<int>> res;
if(!root) return res;
dfs(root,0,res);
return res;
    }
    void dfs(TreeNode* node,int level,vector<vector<int>> &res){
        if(res.size()==level){
            res.push_back(vector<int>{});
        }
        res[level].push_back(node->val);
        if (node->left)dfs(node->left,level+1,res);
        if(node->right)dfs(node->right,level+1,res);
        
    }
};
// @lc code=end

